gpt4all-chat/src/chatllm.cpp

```diff
@@ -693,148 +706,283 @@ void ChatLLM::modelChangeRequested(const ModelInfo &modelInfo)
+        chatTemplate = *tmpl;
+    } else if (chatTemplateSetting.isLegacy()) {
+        throw std::logic_error("cannot apply Jinja to a legacy prompt template");
+    } else {
+        throw std::logic_error("cannot apply Jinja without setting a chat template first");
+    }
+    if (isAllSpace(chatTemplate)) {
+        throw std::logic_error("cannot apply Jinja with a blank chat template");
+    }
+    if (auto tmpl = mySettings->modelSystemMessage(m_modelInfo).asModern()) {
+        systemMessage = *tmpl;
+    } else {
+        throw std::logic_error("cannot apply Jinja with a legacy system message");
+    }
+
+    uint version = parseJinjaTemplateVersion(chatTemplate);
+
+    auto makeMap = [version](const ChatItem &item) {
+        return jinja2::GenericMap([msg = std::make_shared<JinjaMessage>(version, item)] { return msg.get(); });
+    };
+
+    std::unique_ptr<ChatItem> systemItem;
+    bool useSystem = !isAllSpace(systemMessage);
+
+    jinja2::ValuesList messages;
+    messages.reserve(useSystem + items.size());
+    if (useSystem) {
+        systemItem = std::make_unique<ChatItem>(ChatItem::system_tag, systemMessage);
+        messages.emplace_back(makeMap(*systemItem));
+    }
+    for (auto &item : items)
+        messages.emplace_back(makeMap(item));
+
+    jinja2::ValuesMap params {
+        { "messages",              std::move(messages) },
+        { "add_generation_prompt", true                },
+    };
+    for (auto &[name, token] : model->specialTokens())
+        params.emplace(std::move(name), std::move(token));
+
+    std::optional<jinja2::Template> tmpl;
+    auto maybeRendered = loadJinjaTemplate(tmpl, chatTemplate.toStdString())
+        .and_then([&] { return tmpl->RenderAsString(params); });
+    if (!maybeRendered)
+        throw std::runtime_error(fmt::format("Failed to parse chat template: {}", maybeRendered.error().ToString()));
+    return *maybeRendered;
+}
+
+auto ChatLLM::promptInternalChat(const QStringList &enabledCollections, const LLModel::PromptContext &ctx)
+    -> ChatPromptResult
+{
+    Q_ASSERT(isModelLoaded());
+    Q_ASSERT(m_chatModel);
 
     QList<ResultInfo> databaseResults;
     const int retrievalSize = MySettings::globalInstance()->localDocsRetrievalSize();
-    if (!fakeReply && !collectionList.isEmpty()) {
-        emit requestRetrieveFromDB(collectionList, prompt, retrievalSize, &databaseResults); // blocks
-        emit databaseResultsChanged(databaseResults);
+    if (!enabledCollections.isEmpty()) {
+        std::optional<std::pair<int, QString>> query;
+        {
+            // Find the prompt that represents the query. Server chats are flexible and may not have one.
+            auto items = m_chatModel->chatItems(); // holds lock
+            Q_ASSERT(items);
+            auto response = items.cend() - 1;
+            if (auto peer = m_chatModel->getPeerUnlocked(response))
+                query = {*peer - items.cbegin(), (*peer)->value};
```

Comment:
cbegin and cend are giving errors with my g++ version

Reply:
Fixed.